### ISSUE TITLE
Fastify v3 plugin upgrade list

### ISSUE THREAD

**Issue by thomheymann** [fastify/fastify#2217]
### Upgrade info

There are a number of breaking changes which can be found here:
https://github.com/fastify/fastify/blob/899382d682846b07bb706346a1a3f1165d87e369/docs/Migration-Guide-V3.md

Ensure you use latest version of fastify-plugin to get correct types:

While upgrading a plugin ensure that:

- All warnings are fixed (e.g. request.req -> request.raw)
- Update type definitions
- Move type tests to `tsd`
- Update minimum version string
- Add Node v14 to test matrix

### Blocker for release

- [x] fastify-compress - https://github.com/fastify/fastify-compress/pull/107
- [x] fastify-cookie - https://github.com/fastify/fastify-cookie/pull/67
- [x] fastify-cors - https://github.com/fastify/fastify-cors/pull/61
- [x] fastify-cli - https://github.com/fastify/fastify-cli/pull/239
- [x] fastify-helmet - https://github.com/fastify/fastify-helmet/pull/70
- [x] fastify-reply-from - https://github.com/fastify/fastify-reply-from/pull/73
- [x] middie - depends on https://github.com/fastify/middie/pull/36
- [x] point-of-view https://github.com/fastify/point-of-view/pull/154
- [x] under-pressure https://github.com/fastify/under-pressure/pull/50
- [x] fastify-etag - https://github.com/fastify/fastify-etag/pull/14
- [x] fastify-express - https://github.com/fastify/fastify-express/pull/1
- [x] fastify-accepts - https://github.com/fastify/fastify-accepts/pull/55
- [x] fastify-auth (@Ethan-Arrowood) - https://github.com/fastify/fastify-auth/pull/66
- [x] fastify-autoload @frikille
- [x] fastify-basic-auth - https://github.com/fastify/fastify-basic-auth/pull/15
- [x] fastify-bearer-auth (@Ethan-Arrowood) - https://github.com/fastify/fastify-bearer-auth/pull/38
- [x] fastify-plugin - https://github.com/fastify/fastify-plugin/pull/85
- [x] fastify-sensible - https://github.com/fastify/fastify-sensible/pull/22
- [x] fastify-jwt (@Ethan-Arrowood) - https://github.com/fastify/fastify-jwt/pull/86
- [x] fastify-static - https://github.com/fastify/fastify-static/pull/125
- [x] fastify-formbody https://github.com/fastify/fastify-formbody/pull/43
- [x] fastify-swagger https://github.com/fastify/fastify-swagger/pull/222 (to be updated)
- [x] fastify-multipart https://github.com/fastify/fastify-multipart/pull/119

### Follow up / left to community 

- [ ] any-schema-you-like
- [x] aws-lambda-fastify https://github.com/fastify/aws-lambda-fastify/pull/42
- [x] fastify-accepts-serializer
- [x] fastify-caching - https://github.com/fastify/fastify-caching/pull/33
- [x] fastify-circuit-breaker - https://github.com/fastify/fastify-circuit-breaker/pull/16
- [ ] fastify-elasticsearch - https://github.com/fastify/fastify-elasticsearch/pull/28 (possibly not needed at all
- [x] fastify-env - https://github.com/fastify/fastify-env/pull/64
- [x] fastify-flash - https://github.com/fastify/fastify-flash/pull/8
- [x] fastify-http-proxy - https://github.com/fastify/fastify-http-proxy/pull/88
- [x] fastify-kafka - https://github.com/fastify/fastify-kafka/pull/7
- [x] fastify-leveldb - https://github.com/fastify/fastify-leveldb/pull/40 (still needs type definitions)
- [x] fastify-mongodb - https://github.com/fastify/fastify-mongodb/pull/90
- [x] fastify-mysql - https://github.com/fastify/fastify-mysql/pull/32
- [x] fastify-nextjs
- [x] fastify-oauth2 - https://github.com/fastify/fastify-oauth2/pull/44
- [x] fastify-postgres - https://github.com/fastify/fastify-postgres/pull/60
- [x] fastify-rate-limit - https://github.com/fastify/fastify-rate-limit/pull/87
- [x] fastify-redis - https://github.com/fastify/fastify-redis/pull/47
- [ ] fastify-response-validation - https://github.com/fastify/fastify-response-validation/pull/2
- [x] fastify-server-timeout - https://github.com/fastify/fastify-server-timeout/pull/1
- [ ] fastify-sftp
- [ ] fastify-soap-client - https://github.com/fastify/fastify-soap-client/pull/1
- [ ] fastify-starter-codesandbox - https://github.com/fastify/fastify-starter-codesandbox/pull/1
- [x] fastify-url-data
- [x] fastify-websocket - https://github.com/fastify/fastify-websocket/pull/54
- [ ] fastify-zipkin - https://github.com/fastify/fastify-zipkin/pull/4
- [x] fastify-routes - https://github.com/fastify/fastify-routes/pull/12
- [x] fastify-routes-stats - https://github.com/fastify/fastify-routes-stats/pull/18

---

if you want to do do one, please comment on this issue and we can assign you to the plugin

**Comment by mcollina** [comment]
`fastify@3.0.3-alpha.1` is on npm as `next` tag.

**Comment by Ethan-Arrowood** [comment]
Going to take on the collection of auth related plugins starting with fastify-auth.

**Comment by Ethan-Arrowood** [comment]
Are we interested in swapping CI from Greenkeeper to Actions?

**Comment by mcollina** [comment]
I would say that we should converge on Github Actions and Dependabot. Greenkeeper is already disabled org-wide.

**Comment by Ethan-Arrowood** [comment]
I also just realized I confused Greenkeeper with Travis oops 😂But yes okay i'll move CIs to Actions and i'll look at enabling dependabot

**Comment by Ethan-Arrowood** [comment]
I was reading up on TypeScript docs how to best do types for our plugins. Plugins in Fastify are functions so I think its best we follow this template as best as possible: [Template: Function](https://www.typescriptlang.org/v2/docs/handbook/declaration-files/templates/module-function-d-ts.html). Additionally, because these are also plugins we can make use of this tempalte too: [Template: Plugin](https://www.typescriptlang.org/v2/docs/handbook/declaration-files/templates/module-plugin-d-ts.html). I don't think its a hard rule to follow but its a good procedure to keep in mind as we are updating or even writing types for the first time for some of these plugins. For reference I think the `fastify-jwt` types are following these templates as best as possible: https://github.com/fastify/fastify-jwt/pull/86

**Comment by thomheymann** [comment]
> I think its best we follow this template as best as possible: [Template: Function](https://www.typescriptlang.org/v2/docs/handbook/declaration-files/templates/module-function-d-ts.html)

Whats the advantage of declaring a namespace and exporting that using `export = xyz` rather than simply exporting the interfaces you want using ES6 style exports (`export interface xyz`)?

**Comment by Ethan-Arrowood** [comment]
@thomheymann I don't know for certain, but I believe it stems from the expected output of a standard TS project. As in if fastify plugins were written in TS and then we generate a `.d.ts` file it would be shaped like those

**Comment by mcollina** [comment]
If any of you would like to push a release to npm, just tag them as `next`

**Comment by Ethan-Arrowood** [comment]
We still bump major right?

**Comment by mcollina** [comment]
yes

**Comment by Eomm** [comment]
> If any of you would like to push a release to npm, just tag them as `next`

to do it with `releasify` we just use:

```
releasify publish -s major -m --npm-dist-tag next --npm-otp 123 --gh-release-prerelease
```

This will create the **pre** release on github also

**Comment by nm-infy** [comment]
I would like to update fastify-kafka

**Comment by mcollina** [comment]
Go for it!

**Comment by joobisb** [comment]
@mcollina i've updated it for fastify-elasticsearch please have a look 
https://github.com/fastify/fastify-elasticsearch/pull/28

**Comment by mcollina** [comment]
Removing good-first-issue label because it's mostly done.

**Comment by dyanakiev** [comment]
https://github.com/hemerajs/fastify-error-page
This one needs update

**Comment by mcollina** [comment]
I think we are ready to ship v3?

**Comment by jsumners** [comment]
> I think we are ready to ship v3?

Need to get #2308 resolved first.

**Comment by mcollina** [comment]
Second list for updating to rc.4 (which had some unfortunately breaking change on the types):


### Blocker for release

- [x] fastify-compress
- [x] fastify-cookie
- [ ] fastify-cors 
- [ ] fastify-cli 
- [ ] fastify-helmet
- [ ] fastify-reply-from
- [ ] middie
- [ ] point-of-view
- [ ] under-pressure
- [ ] fastify-etag
- [ ] fastify-express
- [ ] fastify-accepts
- [ ] fastify-auth
- [ ] fastify-autoload
- [ ] fastify-basic-auth
- [ ] fastify-bearer-auth 
- [ ] fastify-plugin
- [ ] fastify-sensible 
- [x] fastify-jwt
- [ ] fastify-static
- [ ] fastify-formbody
- [ ] fastify-swagger
- [ ] fastify-multipart 

### Follow up / left to community 

- [ ] any-schema-you-like
- [ ] aws-lambda-fastify
- [ ] fastify-accepts-serializer
- [ ] fastify-caching
- [ ] fastify-circuit-breaker 
- [ ] fastify-elasticsearch
- [ ] fastify-env
- [ ] fastify-flash
- [ ] fastify-http-proxy
- [ ] fastify-kafka
- [ ] fastify-leveldb
- [ ] fastify-mongodb
- [ ] fastify-mysql
- [ ] fastify-nextjs
- [ ] fastify-oauth2 
- [ ] fastify-postgres
- [ ] fastify-rate-limit
- [ ] fastify-redis
- [ ] fastify-response-validation
- [ ] fastify-server-timeout
- [ ] fastify-sftp
- [ ] fastify-soap-client
- [ ] fastify-starter-codesandbox
- [ ] fastify-url-data
- [ ] fastify-websocket
- [ ] fastify-zipkin
- [ ] fastify-routes
- [ ] fastify-routes-stats

**Comment by kibertoad** [comment]
@mcollina I can take some. What is the fastify way to distribute work that reduces chances of effort duplication?

**Comment by mcollina** [comment]
Do a PR on a module and paste the link in this issue :). It should take at most 5-10 minutes to do one.

**Comment by mcollina** [comment]
fastify-nextjs has some errors, it would be great if somebody could check it out.

**Comment by Ethan-Arrowood** [comment]
looking to update fastify-nextjs today.

**Comment by chasingSublimity** [comment]
I think fastify/fastify-compress#113 resolves the `rc.4` incompatibility for `fastify-compress`. Will try and grab a few more tonight. 🚀

**Comment by chasingSublimity** [comment]
So, if `fastify@3.0.0-rc.4` is installed and there aren't any type or test failures, is it reasonable to assume that there are no further compatibility issues? If so, `fastify-compress` and `fastify-mongodb` should be good to go.

**Comment by delvedor** [comment]
Modules that still need to be bumped from next to latest:
Second list for updating to rc.4 (which had some unfortunately breaking change on the types) or just checks that they do not use `FastifyRequestInterface` or `FastifyReplyInterface`:

### Blocker for release

- [ ] fastify-compress
- [ ] fastify-cookie
- [ ] fastify-cors 
- [ ] fastify-cli 
- [ ] fastify-helmet
- [ ] fastify-reply-from - https://github.com/fastify/fastify-reply-from/pull/84
- [ ] middie
- [ ] point-of-view
- [ ] under-pressure
- [ ] fastify-etag
- [ ] fastify-express
- [ ] fastify-accepts
- [ ] fastify-auth
- [ ] fastify-autoload
- [ ] fastify-basic-auth
- [ ] fastify-bearer-auth 
- [ ] fastify-plugin
- [ ] fastify-sensible 
- [ ] fastify-jwt
- [ ] fastify-static
- [ ] fastify-formbody
- [ ] fastify-swagger
- [ ] fastify-multipart 

### Follow up / left to community 

- [ ] any-schema-you-like
- [ ] aws-lambda-fastify
- [ ] fastify-accepts-serializer
- [ ] fastify-caching
- [ ] fastify-circuit-breaker 
- [ ] fastify-elasticsearch
- [ ] fastify-env
- [ ] fastify-flash
- [ ] fastify-http-proxy
- [ ] fastify-kafka
- [ ] fastify-leveldb
- [ ] fastify-mongodb
- [ ] fastify-mysql
- [ ] fastify-nextjs
- [ ] fastify-oauth2 
- [ ] fastify-postgres
- [ ] fastify-rate-limit
- [ ] fastify-redis
- [ ] fastify-response-validation
- [ ] fastify-server-timeout
- [ ] fastify-sftp
- [ ] fastify-soap-client
- [ ] fastify-starter-codesandbox
- [ ] fastify-url-data
- [ ] fastify-websocket
- [ ] fastify-zipkin
- [ ] fastify-routes
- [ ] fastify-routes-stats

**Comment by adrai** [comment]
https://github.com/fastify/aws-lambda-fastify/commit/73cf70b5f247c38b8a20487259d195f4a8bb02b3

**Comment by delvedor** [comment]
Closing this as [v3 is out](https://medium.com/@fastifyjs/fastify-v3-released-da4601d79a5)!
